File: github_code/avl-tree/cpp/66.c
/**
 * Program Name: CS 3610 Project #2 Fall 2017
 * Author: Joshua Wright
 * Date: 10/10/2017
 * Purpose: Implements the AVL Tree data type with nodes of type int.
 * Note: AVL Tree Class Header File
**/

#include <string>
#ifndef AVL_TREE_H
#define AVL_TREE_H


struct BinaryNode {
  BinaryNode(const int key) : key(key), left(NULL), right(NULL) {}

  ~BinaryNode() {
    delete left;
    delete right;
    left = right = NULL;
  }

  int key;
  BinaryNode *left, *right;
  int bfactor = 0;
};

class AVLTree {

public:
  // Constructor
  AVLTree() : root(NULL) {}
  // Deconstructor
  ~AVLTree() { delete root; }

  /* *
   * Function Name: insert
   * Description: Call the private insertion function.
   * Parameters: key - The node key to be inserted
   * */
  void insert(const int key);

  /* *
   * Function Name: remove
   * Description: Delete a key-specified node in the AVL Tree.
   * Parameters: key - The key of the node to be deleted.
   * */
  void remove(const int key);

  /* *
   * Function Name: print
   * Description: Print AVL Tree node keys.
   * Parameters: c - 'p' will print the keys, 'h' will print the (sub)tree height
   * */
  void print(const char c);

  
private:
  /* *
   * Function Name: insert
   * Description: Insert a new node into the AVL-Tree
   * Parameters: key - number to be inserted, node - root of AVL Tree
   * */
  BinaryNode* insert(const int key, BinaryNode* node);

  /* *
   * Function Name: remove
   * Description: Delete a specific node from the AVL-Tree
   * Parameters: key - The key of the node to be deleted, node - The AVL Tree's root node
   * */
  BinaryNode* remove(const int key, BinaryNode* node);

  /* *
   * Function Name: balanceUpdate
   * Description: Update the balance factors of the AVL-Tree's nodes
   * Parameters: node - Root of AVL Tree.
   * */
  void balanceUpdate(BinaryNode* &node);

  /* *
   * Function Name: single_left_rotate
   * Description: Rotate the specified node and its left subtree to the left.
   * Parameters: node - The node to be rotated on.
   * */
  void single_left_rotate(BinaryNode* &node);

  /* *
   * Function Name: single_right_rotate
   * Description: Rotate the specified node and its right subtree to the right.
   * Parameters: node - The node to be rotated on.
   * */
  void single_right_rotate(BinaryNode* &node);

  /* *
   * Function Name: double_left_rotate
   * Description: Rotate the specified node's left child to the left
   *  and the node itself subtree to the right.
   * Parameters: node - The node to be rotated on.
   * */
  void double_left_rotate(BinaryNode* &node);

  /* *
   * Function Name: double_right_rotate
   * Description: Rotate the specified node's left child to the left
   *  and the node itself to the right.
   * Parameters: node - The node to be rotated on.
   * */
  void double_right_rotate(BinaryNode* &node);

  /* *
   * Function Name: inorder
   * Description: Traverse the AVL-Tree in ascending
   * order of node keys, while outputing the key.
   * Parameters: node - Root of the AVL Tree.
   * */
  void inorder(BinaryNode* node);

  /* *
   * Function Name: inorder_height
   * Description: Traverse the AVL-Tree in ascending
   * order of node keys, while outputing the (sub)tree height.
   * Parameters: node - Root of the AVL Tree.
   * */
  void inorder_height(BinaryNode* node);

  /* *
   * Function Name: height
   * Description: Determine the longest path within the AVL-Tree
   * from root to leaf(NULL).
   * Parameters: node - Root of the AVL Tree.
   * */
  int height(BinaryNode* node);

  /* *
   * Function Name: max
   * Description: Determine the greater of two integers.
   * Parameters: i - 1st node's key, j - 2nd node's key
   * */
  int max(int i, int j);

  /* *
   * Function Name: successor
   * Description: Find the next inorder node's key.
   * Parameters: node - The root of the AVL Tree to be traversed.
   * */
  int successor(BinaryNode* node);

  /* *
   * Function Name: minNode
   * Description: Find the node of least value in the AVL Tree.
   * Parameters: node - The root of the AVL Tree where the minimum is wanted.
   * */
  BinaryNode* minNode(BinaryNode* node);


  BinaryNode* root;
};

#endif

File: github_code/avl-tree/java/66.java
package org.pang.learning.ds;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

public class AVL {
	public static void main(String[] args) {
		Tree tree = newTree(2,1,2,5,6,8,9,2,3,12,7,9);
		tree.print();
	}
	
	public static Tree newTree(int ...datas){
		if(datas==null||datas.length<1)
			return null;
		Node root=new Node(datas[0]);
		Tree tree=new Tree(root);
		
		for(int i=1;i<datas.length;i++){
			tree.add(datas[i]);
		}
		
		return tree;
	}

	static class Tree{
		Node root;
		public Tree(Node root){
			this.root=root;
		}
		
		public void add(int data){
			Node node=root;
			do{
				if(node.data==data)
					break;
				if(node.data>data){
					if(node.left==null){
						node.left=new Node(data);
						node.left.parent=node;
						break;
					}else
						node=node.left;
				}else{
					if(node.right==null){
						node.right=new Node(data);
						node.right.parent=node;
						break;
					}else
						node=node.right;
				}
			}while(true);
		}
		
		public int height(){
			return height(root);
		}
		
		private int height(Node node){
			if(node==null)
				return 0;
			else{
				return 1+Math.max(height(node.left), height(node.right));
			}
		}
		
		public void print(Node node){
//			if(node)
		}
		
		public void print(){
			int height=height();
//			System.out.println("height of tree: "+height);
//			System.out.println();
			if(height==0)
				return;
			int distance=2;
			int offset=distance*height;
			LinkedList<Node> nodeList=new LinkedList<Node>();
			nodeList.addLast(root);
			int hIdx=0;
			while(!nodeList.isEmpty()){
				int size=nodeList.size();
				int innerOffset=offset;
				for(Node node : nodeList){
					System.out.printf("%"+innerOffset+"s",node.data);
					innerOffset+=distance*hIdx;
				}
				System.out.println();
				hIdx++;
				offset=offset-distance;
				for(int i=0;i<size;i++){
					Node node = nodeList.poll();
					if(node.left!=null)
						nodeList.addLast(node.left);
					if(node.right!=null)
						nodeList.addLast(node.right);
					
				}
			}
		}
	}
	
	static class Node{
		int data;
		Node left;
		Node right;
		Node parent;
		public Node(int data){
			this.data=data;
		}
	}
}
